cmd/istio-cni-daemonset/main.go

```diff
@@ -0,0 +1,363 @@
+			return
+		}
+	}
+	return
+}
+
+// Delete all pods detected as broken by ListPods
+func (bpr brokenPodReconciler) deleteBrokenPods(filters Filters, options ControllerOptions) error {
+	// Get a list of all broken pods
+	var podList, err = bpr.listBrokenPods(filters, options)
+	if err != nil {
+		return err
+	}
+
+	for _, pod := range podList.Items {
+		log.Infof("Deleting broken pod: %s/%s", pod.Namespace, pod.Name)
+		if err := bpr.client.CoreV1().Pods(pod.Namespace).Delete(pod.Name, nil); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (bpr brokenPodReconciler) listBrokenPods(filters Filters, options ControllerOptions) (list v1.PodList, err error) {
+
+	var rawList *v1.PodList
+	rawList, err = bpr.client.CoreV1().Pods("").List(metav1.ListOptions{
+		TypeMeta:            metav1.TypeMeta{},
+		LabelSelector:       "",
+		FieldSelector:       fmt.Sprintf("%s", filters.FieldSelectors),
+		Watch:               false,
+		AllowWatchBookmarks: false,
+		ResourceVersion:     "",
+		TimeoutSeconds:      nil,
+		Limit:               0,
+		Continue:            "",
+	})
+	if err != nil {
+		return
+	}
+
+	list.Items = []v1.Pod{}
+	for _, pod := range rawList.Items {
+		if detectPod(pod, filters) {
+			list.Items = append(list.Items, pod)
+		}
+	}
+
+	return
+}
+
+// After initial filtering, detect whether this pod is broken due to the CNI
+// race condition.
+func detectPod(pod v1.Pod, filters Filters) bool {
+	// Only check pods that have the sidecar annotation; the rest can be
+	// ignored.
+	if _, ok := pod.ObjectMeta.Annotations["sidecar.istio.io/status"]; ok {
+		// For each candidate pod, iterate across all init containers searching for
+		// crashlooping init containers that match our criteria
+		for _, container := range pod.Status.InitContainerStatuses {
+			// First check that the init container name matches our expected value
+			if container.Name == filters.InitContainerName {
+				// Next check that the container terminated
+				if state := container.LastTerminationState.Terminated; state != nil {
+					// Finally, check that the container state matches our filter criteria
+					// (namely, that the exit code and termination message look right)
+					if checkTerminationState(filters, *state) {
+						return true
+					}
+				}
+			}
+		}
+	}
+	return false
+}
+
+// This func checks the state of a container in a pod and returns true if the container
+// matches the defined filter criteria
+func checkTerminationState(filter Filters, state v1.ContainerStateTerminated) bool {
+	// If we are filtering on init container termination message and the termination message of 'state' does not match, exit
+	if s := strings.TrimSpace(filter.InitContainerTerminationMessage); s != "" && s != strings.TrimSpace(state.Message) {
+		return false
+	}
+
+	// If we are filtering on init container exit code and the termination message does not match, exit
+	if ec := filter.InitContainerExitCode; ec != 0 && ec != int(state.ExitCode) {
```

Comment:
why the extra variable ?

Reply:
```go
return filter.InitContainerExitCode == int(state.ExitCode)
```

Reply:
filter.InitContainerExitCode defaults to zero if unset. We never want to flag a pod as broken if the init container returns  an exit code of zero (successful execution). Without checking for non-zero exit code, that statement would end up returning false positives for pods that have successfully completed their init container and are now running their application payload.

I could change the default ExitCode to something nonsensical, I just left it as zero since that is not an exit code we would ever expect to filter on and so is an easy code to check for. 

As for why I used the 'ec' variable, it was mostly just to shorten the if statement and have it match the pattern set by the InitContainerTerminationMessage if statement above it.